src/rgw/driver/rados/rgw_putobj_processor.cc

```diff
@@ -198,16 +198,16 @@ RadosWriter::~RadosWriter()
    * we remove all the other raw objects. Note that we use different call to remove the head object,
    * as this one needs to go via the bucket index prepare/complete 2-phase commit scheme.
    */
-  for (const auto& obj : written) {
-    if (raw_head && obj == *raw_head) {
-      ldpp_dout(dpp, 5) << "NOTE: we should not process the head object (" << obj << ") here" << dendl;
+  for (auto obj = written.rbegin(); obj != written.rend(); obj++) {
+    if (raw_head && *obj == *raw_head) {
+      ldpp_dout(dpp, 5) << "NOTE: we should not process the head object (" << *obj << ") here" << dendl;
```

Comment:
thanks @thenamehasbeentake

i'm having trouble understanding why this needs to loop in reverse. for multipart uploads, it's the `raw_head` object that we want delete last. we already skip that object in the loop and handle it below

Reply:
ok, i think i understand this now

for multipart uploads, `raw_head` points at the final multipart object name rather than the current part's head object, so this logic doesn't guarantee that the part head object gets deleted last

i uploaded a multipart object `testbucket/128m.iso` with some extra debug logging, and the part uploads show:
> req 14841384554456452370 0.075002573s s3:put_obj raw_head default.rgw.buckets.data:1cf80de1-4e97-4a8f-89d2-747f9b8c432f.4143.1_128m.iso

i pushed a fix in https://github.com/ceph/ceph/pull/55582 and there i see the correct part head:
> req 17160862790303507845 0.070002414s s3:put_obj raw_head default.rgw.buckets.data:8ec2917b-cefa-45a9-b342-3d285fc73685.4143.1__multipart_128m.iso.2\~6HwU84cYMDBfu7p-hOWb4ReeAhCMHRz.2

because of this bug, reversing the order of deletion does help 👍 but i'd rather fix that head_obj detection to stay consistent with the comment history just above